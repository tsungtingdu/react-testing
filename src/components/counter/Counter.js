import React, { useState } from 'react'

const Counter = () => {
  const [counter, setCounter] = useState(0)

  const delayCount = () => {
    setTimeout(() => {
      setCounter(counter + 1)
    }, 1000)
  }

  return (
    <>
      <h1 data-testid="counter">{counter}</h1>
      <button data-testid="button-up" onClick={() => setCounter(counter + 1)}> Up</button>
      <button data-testid="button-down" onClick={() => setCounter(counter - 1)}>Down</button>
      <button data-testid="button-delay" onClick={delayCount}>Down</button>
    </>
  )
}

export default Counter