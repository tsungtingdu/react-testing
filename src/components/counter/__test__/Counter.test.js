import React from 'react'
import { render, cleanup, fireEvent, waitForElement } from '@testing-library/react'
import Counter from './../Counter'

afterEach(cleanup)

// snapshot
it('should take a snapshot', () => {
  const { asFragment } = render(<Counter />)
  expect(asFragment(<Counter />)).toMatchSnapshot()
})

// DOM element
it('should equal to 0', () => {
  const { getByTestId } = render(<Counter />)
  expect(getByTestId('counter')).toHaveTextContent(0)
});

it('should be enabled', () => {
  const { getByTestId } = render(<Counter />)
  expect(getByTestId('button-up')).not.toHaveAttribute('disabled')
})

// DOM event
it('increments counter', () => {
  const { getByTestId } = render(<Counter />)
  fireEvent.click(getByTestId('button-up'))
  expect(getByTestId('counter')).toHaveTextContent(1)
})

it('decrements counter', () => {
  const { getByTestId } = render(<Counter />)
  fireEvent.click(getByTestId('button-down'))
  expect(getByTestId('counter')).toHaveTextContent(-1)
})

// DOM Async Event
it('increments counter after 1s', async () => {
  const { getByTestId, getByText } = render(<Counter />)
  fireEvent.click(getByTestId('button-delay'))
  const counter = await waitForElement(() => getByText('1'))
  expect(counter).toHaveTextContent('1')
})