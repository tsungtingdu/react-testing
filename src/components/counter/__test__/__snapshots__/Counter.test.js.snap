// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <h1
    data-testid="counter"
  >
    0
  </h1>
  <button
    data-testid="button-up"
  >
     Up
  </button>
  <button
    data-testid="button-down"
  >
    Down
  </button>
  <button
    data-testid="button-delay"
  >
    Up (delay)
  </button>
</DocumentFragment>
`;
